extract_metrics.py

```diff
@@ -1,17 +1,225 @@
+    # if we find a function or class name, check for docstrings
+    if node_type not in (get_class_type(py150k) + get_func_type(py150k)):
+        return
+    doc = get_docstring(ast_node, py150k)
+    if not doc:
+        return
+    counter["ds_count"] += 1
+    counter["ds_char_len_total"] += len(doc)
+    counter["ds_word_len_total"] += len(doc.split())
+    counter["ds_line_count"] += len(doc.split("\n"))
+    if node_type in get_func_type(py150k):
+        counter["ds_of_method"] += 1
+
+
+def count_docstring_density(counter):
+    counter["ds_density"] = calculate_ratio(
+        counter["ds_line_count"], counter["line_count"]
+    )
+
+
+def count_docstring_average(counter):
+    counter["ds_char_len_avg"] = calculate_ratio(
+        counter["ds_char_len_total"], counter["ds_count"]
+    )
+    counter["ds_word_len_avg"] = calculate_ratio(
+        counter["ds_word_len_total"], counter["ds_count"]
+    )
+
+
+# Formatting
+# Casing Counts
+def count_casing(ast_node, node_type, counter, py150k=False):
+    # if we find a variable, count its casing
+    if node_type not in (
+        get_var_type(py150k) + get_class_type(py150k) + get_func_type(py150k)
+    ):
+        return
+
+    if py150k:
+        token = ast_node.value
+    elif node_type in get_var_type(py150k):
+        token = ast_node.id
+    else:
+        token = ast_node.name
+
+    case = casing(token)
+
+    counter["id_total"] += 1
+    counter[case] += 1
+
+    if node_type in get_var_type(py150k):
+        counter["id_total_var"] += 1
+        counter[f"{case}_var"] += 1
+    if node_type in get_class_type(py150k):
+        counter["id_total_class"] += 1
+        counter[f"{case}_class"] += 1
+    if node_type in get_func_type(py150k):
+        counter["id_total_method"] += 1
+        counter[f"{case}_method"] += 1
+
+
+def count_casing_ratio(counter):
+    for case in list(CASING_REGEX_MAPPING.keys()) + ["other_case"]:
+        counter[f"{case}_ratio"] = calculate_ratio(
+            counter[case], counter["id_total"]
+        )
+        for id_type in ["var", "class", "method"]:
+            counter[f"{case}_{id_type}_ratio"] = calculate_ratio(
+                counter[f"{case}_{id_type}"], counter[f"id_total_{id_type}"]
+            )
+
+
+# Line Metrics
+def count_lines(code, counter):
+    counter["line_count"] += len(code.split("\n"))
+
+
+# Methods/Classes
+# Method Metrics
+def count_method(ast_node, node_type, counter, py150k=False):
+    # if func def found see how many parents and decorators
+    if node_type not in get_func_type(py150k):
+        return
+    counter["func_count"] += 1
+    counter["func_decorators_count"] += len(get_decorators(ast_node, py150k))
+
+
+def count_async_method(ast_node, node_type, counter, py150k=False):
+    if node_type not in get_async_func_type(py150k):
+        return
+    counter["func_async_count"] += 1
+
+
+def count_method_ratio(counter):
+    counter["func_decotrators_avg"] = calculate_ratio(
```

Comment:
"func_decotrators_avg" should prob be "func_decorators_avg"

Reply:
thanks! fixed